### ISSUE TITLE
`setup.py test --coverage` fails to find source files on python 3 if run on Travis ...

### ISSUE THREAD

**Issue by hamogu** [astropy/package-template#151]
... and thus the report on coverage.io is empty.

I did two runs with identical source and almost identical `.travis.yml` where the only change is that I set a different python version for the test that runs the coverage results:

See these coverage results:
python 2: https://coveralls.io/jobs/12065924
python 3: https://coveralls.io/jobs/12063384

The important part of the Travis logs are here:
python 2: https://travis-ci.org/hamogu/astrospec/jobs/109594431#L536
python 3: https://travis-ci.org/hamogu/astrospec/jobs/109573956#L560

**Comment by astrofrog** [comment]
I see the same issue - investigating

**Comment by astrofrog** [comment]
I think the issue is that the source files get deleted before the coverage is submitted:

```
$ if [[ $SETUP_CMD == 'test --coverage' ]]; then coveralls --rcfile='packagename/tests/coveragerc'; fi
Submitting coverage to coveralls.io...
No source for /tmp/packagename-test-x_re0z2u/lib.linux-x86_64-3.5/packagename/__init__.py
No source for /tmp/packagename-test-x_re0z2u/lib.linux-x86_64-3.5/packagename/example_mod.py
No source for /tmp/packagename-test-x_re0z2u/lib.linux-x86_64-3.5/packagename/example_subpkg/__init__.py
No source for /tmp/packagename-test-x_re0z2u/lib.linux-x86_64-3.5/packagename/extern/__init__.py
No source for /tmp/packagename-test-x_re0z2u/lib.linux-x86_64-3.5/packagename/utils/__init__.py
Coverage submitted!
```

So it must be a difference in behavior between the temporary file mechanism in Python 2 and 3

**Comment by astrofrog** [comment]
Ah, the issue is that normally we patch up the paths to not point to the temp files, but this is not working in Python 3 for some reason.

**Comment by astrofrog** [comment]
Ah well, this would be why:

```
    # The coverage report includes the full path to the temporary
    # directory, so we replace all the paths with the true source
    # path. This means that the coverage line-by-line report will only
    # be correct for Python 2 code (since the Python 3 code will be
    # different in the build directory from the source directory as
    # long as 2to3 is needed). Therefore we only do this fix for
    # Python 2.x.
    if six.PY2:
        try:
            # Coverage 4.0: _harvest_data has been renamed to get_data, the
            # lines dict is private
            cov.get_data()
```

Now that we don't use 2to3 in the core, I'll open a PR.

**Comment by astrofrog** [comment]
Fix in #4822
